File: examples/particleSystems/strings.js

importPackage(com.fishuyo)
importPackage(com.fishuyo.io)
importPackage(com.fishuyo.io.leap)
importPackage(com.fishuyo.maths)
importPackage(com.fishuyo.graphics)
importPackage(com.fishuyo.examples.particleSystems.strings)


// var fabric = Main.fabric()

Mouse.clear()
Mouse.use()
Mouse.bind("drag", function(i){ 
	var x = (i[0] / SimpleAppRun.app().width()) * 20.0
	var y = (1 - i[1] / SimpleAppRun.app().height()) * 20.0

	Main.strings().apply(0).pins().apply(0).position().set(x,y,0)
	
})

Trackpad.connect()
Trackpad.clear()
Trackpad.bind(function(i,f){
	var s=10.0
	if( i == 1){
		Main.strings().apply(0).pins().apply(0).position().lerpTo(Vec3(s*f[5],s*f[6],0), 0.1)
		Main.strings().apply(1).pins().apply(0).position().lerpTo(Vec3(s*f[5],s*f[6],0), 0.1)
		Main.strings().apply(2).pins().apply(0).position().lerpTo(Vec3(s*f[5],s*f[6],0), 0.1)
		Main.strings().apply(3).pins().apply(0).position().lerpTo(Vec3(s*f[5],s*f[6],0), 0.1)
		Main.strings().apply(4).pins().apply(0).position().lerpTo(Vec3(s*f[5],s*f[6],0), 0.1)
	}else if( i == 2){
		Main.strings().apply(0).pins().apply(0).position().lerpTo(Vec3(s*f[5],s*f[6],0), 0.1)
		Main.strings().apply(1).pins().apply(0).position().lerpTo(Vec3(s*f[7],s*f[8],0), 0.1)

	}else if( i==3 ){
		Main.strings().apply(0).pins().apply(0).position().lerpTo(Vec3(s*f[5],s*f[6],0), 0.1)
		Main.strings().apply(1).pins().apply(0).position().lerpTo(Vec3(s*f[7],s*f[8],0), 0.1)
		Main.strings().apply(2).pins().apply(0).position().lerpTo(Vec3(s*f[9],s*f[10],0), 0.1)

	}else if( i==4){
		Main.strings().apply(0).pins().apply(0).position().lerpTo(Vec3(s*f[5],s*f[6],0), 0.1)
		Main.strings().apply(1).pins().apply(0).position().lerpTo(Vec3(s*f[7],s*f[8],0), 0.1)
		Main.strings().apply(2).pins().apply(0).position().lerpTo(Vec3(s*f[9],s*f[10],0), 0.1)
		Main.strings().apply(3).pins().apply(0).position().lerpTo(Vec3(s*f[11],s*f[12],0), 0.1)


	}else if( i==5){
		Main.strings().apply(0).pins().apply(0).position().lerpTo(Vec3(s*f[5],s*f[6],0), 0.1)
		Main.strings().apply(1).pins().apply(0).position().lerpTo(Vec3(s*f[7],s*f[8],0), 0.1)
		Main.strings().apply(2).pins().apply(0).position().lerpTo(Vec3(s*f[9],s*f[10],0), 0.1)
		Main.strings().apply(3).pins().apply(0).position().lerpTo(Vec3(s*f[11],s*f[12],0), 0.1)
		Main.strings().apply(4).pins().apply(0).position().lerpTo(Vec3(s*f[13],s*f[14],0), 0.1)

	}
})

Leap.connect()
Leap.clear()
Leap.bind(function(frame){

	if( frame.hands().isEmpty() ) return
	var hand = frame.hands().get(0)
	if( hand.fingers().isEmpty() ) return
	var count = hand.fingers().count()

	var s = 0.01

	for( i=0; i < count; i++){
		var finger = hand.fingers().get(i)
		var pos = finger.tipPosition()
		Main.strings().apply(i).pins().apply(0).position().lerpTo(Vec3(s*pos.getX(),s*pos.getY(),s*pos.getZ()), 0.1)
	}

})

function step( dt ){
	// println("hi")
	Shader.lightAmbient().set( RGBA.apply(0,1,1,1))

	var p = Main.strings().apply(0).pins().apply(0).position()
	p = Vec3(p.x(),p.y(),p.z()+50)
	// SimpleAppRun.app().camera().nav().pos().lerpTo(p, 0.05)

}

File: examples/particleSystems/strings.scala

package com.fishuyo
package examples.particleSystems.strings

import scala.collection.mutable.ListBuffer

import graphics._
import dynamic._
import maths._
import particle._
import io._

import com.badlogic.gdx.graphics._


object Main extends App with GLAnimatable{

  SimpleAppRun.loadLibs()
  GLScene.push(this)

  val strings = ListBuffer[String]()
  for( i<-(0 until 5)) strings += new String(Vec3(), .5f, .01f, 1.0f)

  val live = new JS("strings.js")

  SimpleAppRun()  

  override def draw(){
  	strings.foreach( _.draw() )
  }

  override def step(dt:Float){
  	strings.foreach( _.step(dt) )
    live.step(dt)
  }

}



class String( var pos:Vec3=Vec3(0), var length:Float=1.f, var dist:Float=.05f, var stiff:Float=1.f) extends GLAnimatable {

  var particles = ListBuffer[Particle]()
  var links = ListBuffer[LinearSpringConstraint]()
  var pins = ListBuffer[AbsoluteConstraint]()

  val numLinks = (length / dist).toInt

  for( i<-(0 to numLinks)){
  	val p = Particle(pos)
  	if( i > 0){
  		links += LinearSpringConstraint(p,particles(i-1),dist,stiff)
  	}
  	particles += p
  }

  pins += AbsoluteConstraint(particles(0), Vec3(pos))
  // pins += AbsoluteConstraint(particles.last, Vec3(pos))

  var vertices = new Array[Float](3*2*numLinks)
  var mesh:Mesh = null

  override def step( dt: Float ) = {

    for( s <- (0 until 5) ){ 
      links.foreach( _.solve() )
      links = links.filter( (l) => !l.isTorn )
      pins.foreach( _.solve() )
    }

    particles.foreach( (p) => {
      p.applyForce(Gravity)
      p.applyDamping(20.f)
      p.step(dt) 
    })

  }

  override def draw() {
    if( mesh == null) mesh = new Mesh(false,2*numLinks,0,VertexAttribute.Position)
    var i = 0
    var off = 0
    for( i<-(0 until links.size)){
	    // if( !links(i).isTorn ){
	      val p = links(i).p.position
	      val q = links(i).q.position
	      val v = i+off
	      vertices(6*v) = p.x
	      vertices(6*v+1) = p.y
	      vertices(6*v+2) = p.z
	      vertices(6*v+3) = q.x
	      vertices(6*v+4) = q.y
	      vertices(6*v+5) = q.z
    	// } else off -= 1
    }
    mesh.setVertices(vertices,0,links.size*6)
    mesh.render( Shader(), GL10.GL_LINES)    
  }

  def applyForce( f: Vec3 ) = particles.foreach( _.applyForce(f) )

}

